presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/db/DbFunctionNamespaceManager.java

```diff
@@ -0,0 +1,285 @@
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static java.util.stream.Collectors.joining;
+
+public class DbFunctionNamespaceManager
+        extends AbstractSqlInvokedFunctionNamespaceManager
+{
+    private static final int MAX_CATALOG_NAME_LENGTH = 128;
+    private static final int MAX_SCHEMA_NAME_LENGTH = 128;
+    private static final int MAX_FUNCTION_NAME_LENGTH = 256;
+    private static final int MAX_PARAMETER_TYPES_LENGTH = 500;
+    private static final int MAX_RETURN_TYPE_LENGTH = 256;
+
+    private final Jdbi jdbi;
+    private final FunctionNamespaceDao functionNamespaceDao;
+
+    @Inject
+    public DbFunctionNamespaceManager(
+            Jdbi jdbi,
+            FunctionNamespaceDao functionNamespaceDao,
+            SqlInvokedFunctionNamespaceManagerConfig managerConfig,
+            DbFunctionNamespaceManagerConfig dbManagerConfig,
+            @ServingCatalog String catalogName)
+    {
+        super(catalogName, managerConfig);
+        this.jdbi = requireNonNull(jdbi, "jdbi is null");
+        this.functionNamespaceDao = requireNonNull(functionNamespaceDao, "functionNamespaceDao is null");
+
+        jdbi.getConfig(SqlStatements.class)
+                .setQueryTimeout(toIntExact(dbManagerConfig.getQueryTimeout().roundTo(SECONDS)));
+        jdbi.getConfig(FunctionNamespacesTableCustomizerFactory.Config.class)
+                .setTableName(dbManagerConfig.getFunctionNamespacesTableName());
+        jdbi.getConfig(SqlFunctionsTableCustomizerFactory.Config.class)
+                .setTableName(dbManagerConfig.getFunctionsTableName());
+    }
+
+    @PostConstruct
+    public void initialize()
+    {
+        functionNamespaceDao.createFunctionNamespacesTableIfNotExists();
+        functionNamespaceDao.createSqlFunctionsTableIfNotExists();
+    }
+
+    @Override
+    public Collection<SqlInvokedFunction> listFunctions()
+    {
+        return functionNamespaceDao.listFunctions(getCatalog());
+    }
+
+    @Override
+    protected Collection<SqlInvokedFunction> fetchFunctionsDirect(QualifiedFunctionName functionName)
+    {
+        checkCatalog(functionName);
+        return functionNamespaceDao.getFunctions(
+                functionName.getFunctionNamespace().getCatalogName(),
+                functionName.getFunctionNamespace().getSchemaName(),
+                functionName.getFunctionName());
+    }
+
+    @Override
+    protected FunctionMetadata fetchFunctionMetadataDirect(SqlFunctionHandle functionHandle)
+    {
+        checkCatalog(functionHandle);
+        Optional<SqlInvokedFunction> function = functionNamespaceDao.getFunction(functionHandle.getFunctionId(), functionHandle.getVersion());
+        if (!function.isPresent()) {
+            throw new InvalidFunctionHandleException(functionHandle);
```

Comment:
Do you want to add an optional message as well?

Reply:
`InvalidFunctionHandleException` adds a error message itself.